mos-platform/common/lib/cxa_abi.cc

```diff
@@ -0,0 +1,157 @@
+ * the advice presented and just initialize the first byte of the 64-bit guard
+ * object.  The remaining 7 bytes are unused (but required to be present by the
+ * ABI standard)
+ */
+
+extern "C" __attribute__((weak)) int __cxa_guard_acquire(uint64_t *guard_object) {
+    // Check if the initializer has run by testing the lowest-address
+    // byte for a non-zero value.
+    return (*reinterpret_cast<uint8_t *>(guard_object) == 0) ? 1 : 0;
+}
+
+extern "C" __attribute__((weak)) void __cxa_guard_release(uint64_t *guard_object) {
+    // set the lowest address byte to a non-zero value.
+    *reinterpret_cast<uint8_t *>(guard_object) = 1;
+}
+
+using FinalizerPtr = void (*)();
+
+static void NoOpFinalize() {}
+
+// Function pointers in .fini_array are invoked from _fini.
+__attribute__((retain, section(".fini_array"))) FinalizerPtr CxaFinalizer = NoOpFinalize;
+
+namespace {
+
+struct ExitFn {
+    void (*m_f)(void *);
+    void * p;
+
+    void operator()() const {
+        m_f(p);
+    }
+};
+
+
+/* Exit functions are registered in a singly-linked list of blocks of registrations.
+ * Each block contains 32 exit registrations, and additional
+ * space for registrations is allocated on the heap, as needed.*/
+class RegistrationList {
+private:
+
+    // FnBlock is an array of function pointers and their arguments
+    // The logical "front" of the block is the last item appended to 
+    // the array.
+    struct FnBlock {
+        static constexpr std::uint8_t BLOCK_SZ = 32;
+
+        ExitFn m_funcs[BLOCK_SZ];
+        std::uint8_t m_sz;
+
+        bool full() const { return m_sz == BLOCK_SZ; }
+
+        void push_front(const ExitFn &newfn) {
+            m_funcs[m_sz++] = newfn;
+        }
+
+        void run_all() {
+            for (std::uint8_t i = m_sz; i > 0; --i) {
+                m_funcs[i-1]();
+            }
+        }
+    };
+
+    struct FnNode : public FnBlock {
+
+        FnNode(FnBlock * next) : m_next{next} {}
+
+        FnBlock * const m_next = nullptr;
+    };
+
+public:
+    static bool push_front(const ExitFn & new_exit) {
+
+        auto & current_block = *m_list;
+
+        if (!current_block.full()) {
+            current_block.push_front(new_exit);
+            return true;
+        }
+        else {
+            const auto next_block = new (std::nothrow) FnNode{m_list};
+            if (!next_block) {
+                // not enough memory to allocate another exit function
+                return false;
+            }
+
+            // link new block to front of list.
+            next_block->push_front(new_exit);
+            m_list = next_block;
+            return true;
+        }
+    }
+
+    static void run_all_exits() {
+        while (m_list != &m_tail) {
+            m_list->run_all();
+
+            const auto current_node_ptr = static_cast<FnNode *>(m_list);
+            const auto next_block = current_node_ptr->m_next;
+            delete current_node_ptr;
```

Comment:
No need; we're about to exit the program. At least until we have DSO unloading, which is an indefinite amount of time in the future.